File: leetcode/divide_two_integers.c
#include <limits.h>

/*
Returns the value of m / n
*/
int divide(int m, int n) {
    int final_ans = 0;
    int sp_ans; // sp = sub-problem
    int ans_sign = -1 * ((m > 0) ^ (n > 0));
    int n_copy;
    
    if (m == INT_MIN && n == -1) {
        return INT_MAX;
    }
    
    if (m != INT_MIN) {
        m = abs(m);    
    }
    n = abs(n);
    

    while (m >= n) {
        sp_ans = 1;
        n_copy = n;
        
        while (m >= n_copy) {
            // Cannot left shift n_copy if it's > 2^30
            if (n_copy > INT_MAX >> 1) {
                break;
            }
            n_copy <<= 1;
            sp_ans <<= 1;
        }

        m -= n_copy >> 1;
        final_ans += sp_ans >> 1;
    }
    
    if (ans_sign == -1) {
        final_ans = ~final_ans + 1; // Or just sign * final_ans
    }

    return final_ans;
}

File: leetcode/java/DivideTwoIntegers.java
class Solution {
    /*
    Returns the value of m / n
    */
    int divide(int m, int n) {
        int finalAns = 0;
        int spAns; // sp = sub-problem
        int ansSign = -1 * ((m > 0 ? 1 : 0) ^ (n > 0 ? 1 : 0));
        int nCopy;
        
        if (m == Integer.MIN_VALUE && n == -1) {
            return Integer.MAX_VALUE;
        }
        
        m = Math.abs(m);
        n = Math.abs(n);
        
        while (m >= n) {
            spAns = 1;
            nCopy = n;
            
            while (m >= nCopy) {
                // Cannot left shift n_copy if it's > 2^30
                if (nCopy > Integer.MAX_VALUE >> 1) {
                    break;
                }
                nCopy <<= 1;
                spAns <<= 1;
            }

            m -= nCopy >> 1;
            finalAns += spAns >> 1;
        }
        
        if (ansSign == -1) {
            finalAns = ~finalAns + 1; // Or just sign * final_ans
        }

        return finalAns;
    }
}